Cast allocas to default address space

Flakebi opened the pull request:
Pointers for variables all need to be in the same address space for correct compilation. Therefore ensure that even if an `alloca` is created in a different address space, it is casted to the default address space before its value is used.

This is necessary for the amdgpu target and others where the default address space for `alloca`s is not 0.

For example the following code compiles incorrectly when not casting the address space to the default one:

```rust
fn f(p: *const i8 /* addrspace(0) */) -> *const i8 /* addrspace(0) */ {
    let local = 0i8; /* addrspace(5) */
    let res = if cond { p } else { &raw const local };
    res
}
```

results in

```llvm
    %local = alloca addrspace(5) i8
    %res = alloca addrspace(5) ptr

if:
    ; Store 64-bit flat pointer
    store ptr %p, ptr addrspace(5) %res

else:
    ; Store 32-bit scratch pointer
    store ptr addrspace(5) %local, ptr addrspace(5) %res

ret:
    ; Load and return 64-bit flat pointer
    %res.load = load ptr, ptr addrspace(5) %res
    ret ptr %res.load
```

For amdgpu, `addrspace(0)` are 64-bit pointers, `addrspace(5)` are 32-bit pointers.
The above code may store a 32-bit pointer and read it back as a 64-bit pointer, which is obviously wrong and cannot work. Instead, we need to `addrspacecast %local to ptr addrspace(0)`, then we store and load the correct type.

Tracking issue: #135024

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

compiler-errors commented:
Isn't this going to add a redundant addrspace cast for LLVM targets that don't care about this?

Flakebi commented:
> Isn't this going to add a redundant addrspace cast for LLVM targets that don't care about this?

No, LLVM checks itself if the cast is needed or not. So, if we checked this here, we would check twice in the case the cast is needed.

The code in LLVM is here:
https://github.com/llvm/llvm-project/blob/dd30aa83aa12e5b2b5e58cb72ec85070f725df34/llvm/include/llvm/IR/IRBuilder.h#L2201-L2204

The code for constant casts is here:
https://github.com/llvm/llvm-project/blob/dd30aa83aa12e5b2b5e58cb72ec85070f725df34/llvm/lib/IR/Instructions.cpp#L3041-L3042
(The `Create(Instruction::BitCast, …)` below is also a no-op if the types are already equal.)

compiler-errors commented:
Maybe r? nikic

Or reassign

nikic reviewed (commented):
LGTM, but I think we should have a codegen test for this, which is presumably only possible after the amdgpu target has landed?

Flakebi commented:
> LGTM, but I think we should have a codegen test for this, which is presumably only possible after the amdgpu target has landed?

Yes, I don’t think I can add one otherwise. Rust rejects datalayout that do not match the LLVM one, so I can’t create a custom target file that has a non-0 default addrspace for `alloca`s.

Flakebi commented:
For what it’s worth, I wrote a test here: https://github.com/Flakebi/rust/commit/38de6b7e14c253de070da8e5eebeca19561e2f10
Merging the amdgpu target seems to take a while longer because some CI jobs run out of disk space with the additional LLVM target.

Flakebi commented:
The target is merged, so the test should pass now 🥳
[Diff](https://github.com/rust-lang/rust/compare/68daddf541253f143ed8359d5c5a9c2dce6abd70..dfd5addb4167312ce2960a09cbf2448bcc7f83f0) (just added the test)

nikic commented on tests/codegen/amdgpu-addrspacecast.rs:
```diff
@@ -0,0 +1,22 @@
+#[lang = "freeze"]
+trait Freeze {}
+#[lang = "copy"]
+trait Copy {}
```
Use `//@ add-core-stubs` instead?

Flakebi committed:
Cast allocas to default address space

Pointers for variables all need to be in the same address space for
correct compilation. Therefore ensure that even if an `alloca` is
created in a different address space, it is casted to the default
address space before its value is used.

This is necessary for the amdgpu target and others where the default
address space for `alloca`s is not 0.

For example the following code compiles incorrectly when not casting the
address space to the default one:

```rust
fn f(p: *const i8 /* addrspace(0) */) -> *const i8 /* addrspace(0) */ {
    let local = 0i8; /* addrspace(5) */
    let res = if cond { p } else { &raw const local };
    res
}
```

results in

```llvm
    %local = alloca addrspace(5) i8
    %res = alloca addrspace(5) ptr

if:
    ; Store 64-bit flat pointer
    store ptr %p, ptr addrspace(5) %res

else:
    ; Store 32-bit scratch pointer
    store ptr addrspace(5) %local, ptr addrspace(5) %res

ret:
    ; Load and return 64-bit flat pointer
    %res.load = load ptr, ptr addrspace(5) %res
    ret ptr %res.load
```

For amdgpu, `addrspace(0)` are 64-bit pointers, `addrspace(5)` are
32-bit pointers.
The above code may store a 32-bit pointer and read it back as a 64-bit
pointer, which is obviously wrong and cannot work. Instead, we need to
`addrspacecast %local to ptr addrspace(0)`, then we store and load the
correct type.

Flakebi commented:
Now using `add-core-stubs` in the test. [Diff](https://github.com/rust-lang/rust/compare/dfd5addb4167312ce2960a09cbf2448bcc7f83f0..cde7e805ad48e8766a07c95c2046693c6a6e236c)

nikic commented:
@bors r+

bors commented:
:pushpin: Commit cde7e805ad48e8766a07c95c2046693c6a6e236c has been approved by `nikic`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=nikic cde7e805ad48e8766a07c95c2046693c6a6e236c -->
<!-- homu: {"type":"Approved","sha":"cde7e805ad48e8766a07c95c2046693c6a6e236c","approver":"nikic","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
